headers.go

```diff
@@ -0,0 +1,130 @@
+package cose
+
+import "github.com/fxamacker/cbor/v2"
+
+// COSE Header labels registered in the IANA "COSE Header Parameters" registry.
+//
+// Reference: https://www.iana.org/assignments/cose/cose.xhtml#header-parameters
+const (
+	HeaderLabelAlgorithm         int64 = 1
+	HeaderLabelCritical          int64 = 2
+	HeaderLabelContentType       int64 = 3
+	HeaderLabelKeyID             int64 = 4
+	HeaderLabelCounterSignature  int64 = 7
+	HeaderLabelCounterSignature0 int64 = 9
+	HeaderLabelX5Bag             int64 = 32
+	HeaderLabelX5Chain           int64 = 33
+	HeaderLabelX5T               int64 = 34
+	HeaderLabelX5U               int64 = 35
+)
+
+// ProtectedHeader contains parameters that are to be cryptographically
+// protected.
+type ProtectedHeader map[interface{}]interface{}
+
+// MarshalCBOR encodes the protected header into a CBOR bstr object.
+// A zero-length header is encoded as a zero-length string rather than as a
+// zero-length map (encoded as h'a0').
+func (h ProtectedHeader) MarshalCBOR() ([]byte, error) {
+	var encoded []byte
+	if len(h) == 0 {
+		encoded = []byte{}
+	} else {
+		var err error
+		encoded, err = encMode.Marshal(map[interface{}]interface{}(h))
+		if err != nil {
+			return nil, err
+		}
+	}
+	return encMode.Marshal(encoded)
+}
+
+// UnmarshalCBOR decodes a CBOR bstr object into ProtectedHeader.
+func (h *ProtectedHeader) UnmarshalCBOR(data []byte) error {
+	var encoded []byte
+	if err := decMode.Unmarshal(data, &encoded); err != nil {
+		return err
+	}
+	if len(encoded) == 0 {
+		(*h) = make(ProtectedHeader)
+	} else {
+		var header map[interface{}]interface{}
+		if err := decMode.Unmarshal(encoded, &header); err != nil {
+			return err
+		}
+		(*h) = header
+	}
+	return nil
+}
+
+// SetAlgorithm sets the algorithm value to the algorithm header.
+func (h ProtectedHeader) SetAlgorithm(alg Algorithm) {
+	h[HeaderLabelAlgorithm] = int64(alg)
+}
+
+// Algorithm gets the algorithm value from the algorithm header.
+func (h ProtectedHeader) Algorithm() (Algorithm, error) {
+	value, ok := h[HeaderLabelAlgorithm]
+	if !ok {
+		return 0, ErrAlgorithmNotFound
+	}
+	alg, ok := value.(int64)

```

Comment:
`value` won't be a `int64`, but a `cose.Algorithm`, if the algorithm has been set like this:

```go
headers := cose.Headers{
		Protected: cose.ProtectedHeader{
			cose.HeaderLabelAlgorithm: cose.AlgorithmES512,
		},
	}
```

I've just realized that with the current `Algorithm` definition it will be hard to not mix `int64` and `Algorithm` here and there. Another example is `ProtectedHeader.SetAlgorithm`, which castes the algorithm to a `int64`. And probably when unmarshalling a header we will also have an `int64`.

To solve this problem, what if we change the `Algorithm` definition to a be an alias instead of a type, so we always deal with `int64` types:

```go
type Algorithm = int64
```

The drawback is that we lose the `String` method, but it was not really being used anyway.

Reply:
We cannot do `type Algorithm = int64`. Otherwise, we are not able to define any methods for this type.

Instead, I store `Algorithm` directly in the map (see 06233d7), and it will be marshaled to int64 anyway.

Reply:
Great. Then I would also change `ProtectedHeader.UnmarshalCBOR` so it also cast the algorithm entry (key=1) into an `Algorithm` if it is present.